tests/integration/ha/test_ha.py

```diff
@@ -107,37 +115,200 @@ async def test_kill_broker_with_topic_leader(
+
+    logger.info(
+        f"Freezing broker of leader for topic '{ContinuousWrites.TOPIC_NAME}': {initial_leader_num}"
+    )
+    await send_control_signal(
+        ops_test=ops_test, unit_name=f"{APP_NAME}/{initial_leader_num}", signal="SIGSTOP"
+    )
+    await asyncio.sleep(25)
+
+    # verify replica is not in sync
+    topic_description = await get_topic_description(
+        ops_test=ops_test, topic=ContinuousWrites.TOPIC_NAME
     )
-    res = c_writes.stop()
+    assert topic_description.in_sync_replicas == {0, 1, 2} - {initial_leader_num}
+    assert not is_up(
+        ops_test=ops_test, broker_id=initial_leader_num
+    ), f"Broker {initial_leader_num} reported as up"
+
+    # verify new writes are continuing by counting the number of writes before and after 5 seconds.
+    # Also, check that leader changed
+    topic_description = await get_topic_description(
+        ops_test=ops_test, topic=ContinuousWrites.TOPIC_NAME
+    )
+    next_leader_num = topic_description.leader
+    initial_offsets = await get_topic_offsets(ops_test=ops_test, topic=ContinuousWrites.TOPIC_NAME)
+    await asyncio.sleep(5)
+    next_offsets = await get_topic_offsets(ops_test=ops_test, topic=ContinuousWrites.TOPIC_NAME)
 
     assert initial_leader_num != next_leader_num
     assert int(next_offsets[-1]) > int(initial_offsets[-1])
-    assert res.lost_messages == "0"
-    assert res.count == int(res.last_expected_message)
+
+    # Un-freeze the process
+    logger.info(f"Un-freezing broker: {initial_leader_num}")
+    await send_control_signal(
+        ops_test=ops_test, unit_name=f"{APP_NAME}/{initial_leader_num}", signal="SIGCONT"
+    )
+    await asyncio.sleep(15)
+    topic_description = await get_topic_description(
+        ops_test=ops_test, topic=ContinuousWrites.TOPIC_NAME
+    )
+
+    # verify the unit is now rejoined the cluster
+    assert is_up(
+        ops_test=ops_test, broker_id=initial_leader_num
+    ), f"Broker {initial_leader_num} reported as down"
+    assert topic_description.in_sync_replicas == {0, 1, 2}
+
+    result = c_writes.stop()
+    assert_continuous_writes_consistency(result=result)
+
+
+async def test_full_cluster_crash(
+    ops_test: OpsTest,
+    c_writes: ContinuousWrites,
+    c_writes_runner: ContinuousWrites,
+    restart_delay,
+):
+    # Let some time pass for messages to be produced
+    await asyncio.sleep(10)
+
+    logger.info("Killing all brokers...")
+    # kill all units "simultaneously"
+    await asyncio.gather(
+        *[
+            send_control_signal(ops_test, unit.name, signal="SIGKILL")
+            for unit in ops_test.model.applications[APP_NAME].units
+        ]
+    )
+    # Give time for the service to restart
+    await asyncio.sleep(RESTART_DELAY * 2)
+
+    topic_description = await get_topic_description(
+        ops_test=ops_test, topic=ContinuousWrites.TOPIC_NAME
+    )
+    initial_offsets = await get_topic_offsets(ops_test=ops_test, topic=ContinuousWrites.TOPIC_NAME)
```

Comment:
**question** shouldn't this be at the top, before killing the brokers? Probably it is the same, but I believe it would make the test a bit clearer, in the sense that we 
1. Set the stage and get pre-information
2. Do the action (killing everything)
3. do asserts

Reply:
Initial offsets being here is intentional, as I want to check that offsets are still increasing _after_ all brokers are back up. If I initialize before killing brokers, the offsets will still report an increase because of the time it takes to actually kill the brokers.